rllib/agents/dqn/learner_thread.py

```diff
@@ -54,22 +84,115 @@ def step(self):
+                accordingly.
+            num_data_load_threads (int): Number of threads to use to load
+                data into GPU memory in parallel.
+        """
+        super().__init__(local_worker=local_worker)
+
+        self.train_batch_size = train_batch_size
+
+        # TODO: (sven) Allow multi-GPU to work for multi-agent as well.
+        self.policy = self.local_worker.policy_map[DEFAULT_POLICY_ID]
+        self.policy_map = self.local_worker.policy_map
+        self.devices = self.policy.devices
+
+        logger.info("APEXMultiGPULearnerThread devices {}".format(self.devices))
+        assert self.train_batch_size % len(self.devices) == 0
+        assert self.train_batch_size >= len(self.devices), "batch too small"
+
+        if set(self.local_worker.policy_map.keys()) != {DEFAULT_POLICY_ID}:
+            raise NotImplementedError("Multi-gpu mode for multi-agent")
+
+        self.tower_stack_indices = list(range(num_multi_gpu_tower_stacks))
+
+        self.idle_tower_stacks = queue.Queue()
+        self.ready_tower_stacks = queue.Queue()
+        for idx in self.tower_stack_indices:
+            self.idle_tower_stacks.put(idx)
+        for i in range(num_data_load_threads):
+            self.loader_thread = _MultiGPULoaderThread(self, share_stats=(i == 0))
+            self.loader_thread.start()
+
+    @override(APEXLearnerThread)
+    def step(self) -> None:
+        assert self.loader_thread.is_alive()
+        with self.load_wait_timer:
+            replay_actor, buffer_idx, batch_indexes = self.ready_tower_stacks.get()
+
+        get_num_samples_loaded_into_buffer = 0
+        with self.grad_timer:
+            # Use LearnerInfoBuilder as a unified way to build the final
+            # results dict from `learn_on_loaded_batch` call(s).
+            # This makes sure results dicts always have the same structure
+            # no matter the setup (multi-GPU, multi-agent, minibatch SGD,
+            # tf vs torch).
+            learner_info_builder = LearnerInfoBuilder(num_devices=len(self.devices))
+
+            default_policy_results = self.policy.learn_on_loaded_batch(
+                offset=0, buffer_index=buffer_idx
+            )
+            learner_info_builder.add_learn_on_batch_results(default_policy_results)
             self.weights_updated = True
-            self.overall_timer.push_units_processed(replay and replay.count or 0)
+            get_num_samples_loaded_into_buffer += (
+                self.policy.get_num_samples_loaded_into_buffer(buffer_idx)
+            )
+
+            self.learner_info = learner_info_builder.finalize()
+
+            td_error = self.learner_info["default_policy"]["td_error"]
+            prio_dict = {"default_policy": (batch_indexes, td_error)}
+
+        self.idle_tower_stacks.put(buffer_idx)
+        self.outqueue.put((replay_actor, prio_dict, get_num_samples_loaded_into_buffer))
+        self.learner_queue_size.push(self.inqueue.qsize())
```

Comment:
a random question, do we need to care about multi-threading related issues with these learner threads? not sure if there are more than 1 such threads actually.